File: sdk/typescript/src/memory/tokens.ts
/**
 * Token estimation utilities for session compaction.
 *
 * Uses a simple heuristic: ~4 characters per token.
 */

import type { ConversationMessage } from '../runtime/orchestrator-types.js';

/**
 * Estimate token count for a string using the ~4 chars/token heuristic.
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

/**
 * Estimate token count for a single conversation message.
 */
export function estimateMessageTokens(message: ConversationMessage): number {
  const content = message.content;
  if (typeof content === 'string') {
    return estimateTokens(content);
  }
  try {
    return estimateTokens(JSON.stringify(content));
  } catch {
    return 0;
  }
}

/**
 * Estimate total token count for an array of conversation messages.
 */
export function estimateMessagesTokens(messages: ConversationMessage[]): number {
  let total = 0;
  for (const msg of messages) {
    total += estimateMessageTokens(msg);
  }
  return total;
}

File: sdk/python/polos/memory/tokens.py
"""Token estimation utilities for session compaction.

Uses a simple heuristic: ~4 characters per token.
"""

from __future__ import annotations

import json
import math


def estimate_tokens(text: str) -> int:
    """Estimate token count for a string using the ~4 chars/token heuristic."""
    return math.ceil(len(text) / 4)


def estimate_message_tokens(message: dict) -> int:
    """Estimate token count for a single conversation message."""
    content = message.get("content")
    if isinstance(content, str):
        return estimate_tokens(content)
    try:
        return estimate_tokens(json.dumps(content))
    except (TypeError, ValueError):
        return 0


def estimate_messages_tokens(messages: list[dict]) -> int:
    """Estimate total token count for an array of conversation messages."""
    total = 0
    for msg in messages:
        total += estimate_message_tokens(msg)
    return total
